File: 10_22_2016_validate-binary-search-tree/chang.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: True if the binary tree is BST, or false
     */
    public boolean isValidBST(TreeNode root) {
        ArrayList<Integer> list = new ArrayList<Integer>();
        inOrderT(root, list);
        
        for (int i = 1; i < list.size(); i++) {
            if (list.get(i-1) >= list.get(i)) {
                return false;
            }
        }
        return true;


    }
    
    private void inOrderT(TreeNode root, ArrayList<Integer> list) {
        //special case
        if (root == null) {
            return;
        }
        //left
        if (root.left != null) {
            inOrderT(root.left, list);
        }
        
        list.add(root.val);
        
        if (root.right != null) {
            inOrderT(root.right, list);
        }
        
        
    }
}

File: 2016-11-28-binary-tree-maximum-path-sum/chang.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """
    def maxPathSum(self, root):
        maxAnyToAny, _ = self.maxPathHelper(root)
        return maxAnyToAny
        
    def maxPathHelper(self, root):
        if root is None:
            # return anyToAny(contains at least one node) and rootToAny 
            return -sys.maxint, 0 
            
        left = self.maxPathHelper(root.left)
        right = self.maxPathHelper(root.right)
        
        rootToAny = max(0, root.val, root.val + left[1], root.val + right[1])
        anyToAny = max(left[0], right[0], left[1] + right[1] + root.val)
        
        return anyToAny, rootToAny
